File: Go/4kyu/4-by-4-skyscrapers.go
// Title: 4 By 4 Skyscrapers
// Link: https://www.codewars.com/kata/5671d975d81d6c1c87000022
// Difficulty: 4 kyu
//
// ## Description
//
// Solve a 4×4 skyscraper puzzle: place heights 1–4 in each cell so every row and column is a permutation of 1–4.
// Clues around the grid (16 values, clockwise; 0 means unknown) give how many buildings are visible from that
// direction; taller buildings hide shorter ones behind them.

package kata

const size = 4

func countVisibleAlong(heights []int, reverse bool) int {
	tallest, seen := 0, 0
	n := len(heights)
	if n == 0 {
		return 0
	}
	idx, step := 0, 1
	if reverse {
		idx = n - 1
		step = -1
	}
	for i := 0; i < n; i++ {
		height := heights[idx]
		if height > tallest {
			tallest = height
			seen++
		}
		idx += step
	}
	return seen
}

func countVisible(heights []int) int {
	return countVisibleAlong(heights, false)
}

func countVisibleFromEnd(heights [4]int) int {
	return countVisibleAlong(heights[:], true)
}

var rowPermutations [][4]int

func init() {
	digits := [4]int{1, 2, 3, 4}
	var partial [4]int
	var generate func(int)
	generate = func(pos int) {
		if pos == size {
			snap := partial
			rowPermutations = append(rowPermutations, snap)
			return
		}
		for _, height := range digits {
			taken := false
			for j := 0; j < pos; j++ {
				if partial[j] == height {
					taken = true
					break
				}
			}
			if taken {
				continue
			}
			partial[pos] = height
			generate(pos + 1)
		}
	}
	generate(0)
}

func rowFitsEdgeClues(candidate [4]int, clueLeft, clueRight int) bool {
	leftOk := clueLeft == 0 || countVisible(candidate[:]) == clueLeft
	rightOk := clueRight == 0 || countVisibleFromEnd(candidate) == clueRight
	return leftOk && rightOk
}

func rowOptionsForIndex(row int, clues []int) [][4]int {
	clueLeft := clues[15-row]
	clueRight := clues[4+row]
	var options [][4]int
	for _, candidate := range rowPermutations {
		if rowFitsEdgeClues(candidate, clueLeft, clueRight) {
			options = append(options, candidate)
		}
	}
	return options
}

func columnFitsEdgeClues(column [4]int, clueTop, clueBottom int) bool {
	topOk := clueTop == 0 || countVisible(column[:]) == clueTop
	bottomOk := clueBottom == 0 || countVisibleFromEnd(column) == clueBottom
	return topOk && bottomOk
}

func columnsMatchTopBottom(grid *[4][4]int, clues []int) bool {
	var col [4]int
	for c := 0; c < size; c++ {
		for r := 0; r < size; r++ {
			col[r] = grid[r][c]
		}
		if !columnFitsEdgeClues(col, clues[c], clues[8+(3-c)]) {
			return false
		}
	}
	return true
}

func columnClashesWithAbove(grid *[4][4]int, row int, choice [4]int) bool {
	for c := 0; c < size; c++ {
		h := choice[c]
		for above := 0; above < row; above++ {
			if grid[above][c] == h {
				return true
			}
		}
	}
	return false
}

func dfs(row int, grid *[4][4]int, rowsByDepth [][][4]int, clues []int) bool {
	if row == size {
		return columnsMatchTopBottom(grid, clues)
	}
	for _, choice := range rowsByDepth[row] {
		if columnClashesWithAbove(grid, row, choice) {
			continue
		}
		for c := 0; c < size; c++ {
			grid[row][c] = choice[c]
		}
		if dfs(row+1, grid, rowsByDepth, clues) {
			return true
		}
	}
	return false
}

func gridToResult(g *[4][4]int) [][]int {
	backing := make([]int, size*size)
	out := make([][]int, size)
	for r := 0; r < size; r++ {
		slice := backing[r*size : (r+1)*size]
		copy(slice, g[r][:])
		out[r] = slice
	}
	return out
}

func SolvePuzzle(clues []int) [][]int {
	if len(clues) != 16 {
		return nil
	}
	rowsByDepth := make([][][4]int, size)
	for i := 0; i < size; i++ {
		rowsByDepth[i] = rowOptionsForIndex(i, clues)
		if len(rowsByDepth[i]) == 0 {
			return nil
		}
	}
	var grid [4][4]int
	if !dfs(0, &grid, rowsByDepth, clues) {
		return nil
	}
	return gridToResult(&grid)
}

File: JavaScript/4kyu/4-by-4-skyscrapers.js
/**
 * Title: 4 By 4 Skyscrapers
 * Link: https://www.codewars.com/kata/5671d975d81d6c1c87000022
 * Difficulty: 4 kyu
 *
 * ## Description
 *
 * Solve a 4×4 skyscraper puzzle: place heights 1–4 in each cell so every row and column is a permutation of 1–4.
 * Clues around the grid (16 values, clockwise; `0` means unknown) give how many buildings are visible from that
 * direction; taller buildings hide shorter ones behind them.
 */

const SIZE = 4;

function countVisible(heights) {
  let tallest = 0;
  let seen = 0;
  for (let i = 0; i < heights.length; i++) {
    const h = heights[i];
    if (h <= tallest) continue;
    tallest = h;
    seen++;
  }
  return seen;
}

function countVisibleFromEnd(heights) {
  return countVisible([heights[3], heights[2], heights[1], heights[0]]);
}

const ROW_PERMUTATIONS = (function buildPermutations1234() {
  const digits = [1, 2, 3, 4];
  const out = [];
  function build(chosen, remaining) {
    if (remaining.length === 0) {
      out.push(chosen.slice());
      return;
    }
    for (let i = 0; i < remaining.length; i++) {
      chosen.push(remaining[i]);
      build(chosen, remaining.filter((_, j) => j !== i));
      chosen.pop();
    }
  }
  build([], digits);
  return out;
})();

function rowFitsEdgeClues(candidate, clueLeft, clueRight) {
  const leftOk = clueLeft === 0 || countVisible(candidate) === clueLeft;
  const rightOk = clueRight === 0 || countVisibleFromEnd(candidate) === clueRight;
  return leftOk && rightOk;
}

function rowOptionsForIndex(rowIndex, clues) {
  const clueLeft = clues[15 - rowIndex];
  const clueRight = clues[4 + rowIndex];
  return ROW_PERMUTATIONS.filter((c) => rowFitsEdgeClues(c, clueLeft, clueRight));
}

function columnFitsEdgeClues(column, clueTop, clueBottom) {
  const topOk = clueTop === 0 || countVisible(column) === clueTop;
  const bottomOk = clueBottom === 0 || countVisibleFromEnd(column) === clueBottom;
  return topOk && bottomOk;
}

function columnsMatchTopBottom(board, clues) {
  for (let col = 0; col < SIZE; col++) {
    const column = [board[0][col], board[1][col], board[2][col], board[3][col]];
    const clueTop = clues[col];
    const clueBottom = clues[8 + (3 - col)];
    if (!columnFitsEdgeClues(column, clueTop, clueBottom)) return false;
  }
  return true;
}

function columnClashesWithAbove(grid, row, choice) {
  for (let col = 0; col < SIZE; col++) {
    const height = choice[col];
    for (let above = 0; above < row; above++) {
      if (grid[above][col] === height) return true;
    }
  }
  return false;
}

function assignRow(grid, row, choice) {
  for (let col = 0; col < SIZE; col++) grid[row][col] = choice[col];
}

function cloneRows(grid) {
  return grid.map((line) => line.slice());
}

function dfs(row, grid, rowsByDepth, clues) {
  if (row === SIZE) {
    return columnsMatchTopBottom(grid, clues) ? cloneRows(grid) : null;
  }
  for (const choice of rowsByDepth[row]) {
    if (columnClashesWithAbove(grid, row, choice)) continue;
    assignRow(grid, row, choice);
    const solved = dfs(row + 1, grid, rowsByDepth, clues);
    if (solved) return solved;
  }
  return null;
}

function solvePuzzle (clues) {
  const rowsByDepth = [0, 1, 2, 3].map((i) => rowOptionsForIndex(i, clues));
  const grid = [
    [0, 0, 0, 0],
    [0, 0, 0, 0],
    [0, 0, 0, 0],
    [0, 0, 0, 0],
  ];
  return dfs(0, grid, rowsByDepth, clues);
}
